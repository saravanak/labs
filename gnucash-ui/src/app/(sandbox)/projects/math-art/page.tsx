import ArtCanvas from '@/components/math-art/art-canvas';

export default async function MathArtPage() {
  // const data = await getData();
  return <ArtCanvas />;
}
