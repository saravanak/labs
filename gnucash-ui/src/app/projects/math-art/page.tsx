import ArtCanvas from "@/components/math-art/art-canvas";

async function getData() {
  const res: any = await fetch("http://localhost:3001/script", {
    method: "POST",
    body: `
export function logic(x,y) {
  return x == 3 ? "blue" : "black" ;
};
`,
  });

  return res.json();
}

export default async function MathArtPage() {
  const data = await getData();
  return <ArtCanvas data={data} />;
}

