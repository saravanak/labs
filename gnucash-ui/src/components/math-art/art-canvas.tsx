'use client';

import { useCallback, useEffect, useRef, useState } from 'react';
import canvasSketch from 'canvas-sketch';
import { EditorView, basicSetup } from 'codemirror';
import { javascript } from '@codemirror/lang-javascript';
import { EditorState } from '@codemirror/state';
import { ViewUpdate, keymap } from '@codemirror/view';
import { defaultKeymap } from '@codemirror/commands';
import readOnlyRangesExtension from 'codemirror-readonly-ranges';

// Specify some output parameters
const settings: any = {
  // The [ width, height ] of the artwork in pixels
  dimensions: [120, 120],
};

const getReadOnlyRanges = (
  targetState: EditorState
): Array<{ from: number | undefined; to: number | undefined }> => {
  return [
    {
      from: targetState.doc.line(1).from,
      to: targetState.doc.line(3).to,
    },
    {
      from: targetState.doc.line(targetState.doc.lines - 2).from,
      to: targetState.doc.line(targetState.doc.lines).to,
    },
  ];
};

export default function ArtCanvas({ data }: any) {
  const canvasRef = useRef(null);
  const codeEditorRef = useRef(null);
  const [isClient, setIsClient] = useState(false);
  const [userCode, setUserCode] = useState(`
  export function logic(x,y) {
    //Add your logic here...
    return x == 3 ? "blue" : "black" ;
    //End your logic here...
  };
  `);

  const sketch = useCallback(
    (props: any) => {
      let expendedArray = [];

      let rle = data.stdout.rle;

      const regexp = new RegExp(/^([bB]\d+)/);

      let match = rle.match(regexp);

      while (match) {
        const [letter, ...count] = match[1].split('');

        expendedArray.push(...Array(Number(count.join(''))).fill(letter));
        rle = rle.slice(match[1].length);
        match = rle.match(regexp);
      }

      expendedArray = expendedArray.map((v) => (v == 'b' ? 'gray' : 'blue'));

      // Destructure what we need from props
      const { context, width, height } = props;

      // Fill the canvas with pink

      for (var i = 0; i < 100; i++) {
        context.fillStyle = expendedArray[i];
        context.fillRect(10 * (i % 10), 10 * Math.floor(i / 10), 10, 10);
      }
    },
    [data?.stdout?.rle]
  );

  useEffect(() => {
    if (codeEditorRef?.current) {
      console.log('instantiating codemirror');

      let startState = EditorState.create({
        doc: userCode,
        extensions: [
          keymap.of(defaultKeymap),
          EditorView.updateListener.of((update: ViewUpdate) => {
            setUserCode((update.state.doc as any).text.join('\n'));
          }),
          readOnlyRangesExtension(getReadOnlyRanges),
        ],
      });

      let view = new EditorView({
        state: startState,
        extensions: [basicSetup, javascript()],
        parent: codeEditorRef.current,
      });
    }
  }, [codeEditorRef, isClient]);

  useEffect(() => {
    setIsClient(true);

    if (canvasRef.current) {
      settings.canvas = canvasRef.current;
      canvasSketch(sketch, {
        ...settings,
        canvas: canvasRef.current,
      });
    }
  }, [canvasRef, isClient, sketch]);
  return (
    <>
      {isClient ? (
        <div>
          <div className='w-[500px] h-[500px]' ref={codeEditorRef} />
          <canvas ref={canvasRef} />
        </div>
      ) : null}
    </>
  );
}
