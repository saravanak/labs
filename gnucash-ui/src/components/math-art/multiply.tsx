"use client";
import { cn } from "@/lib/utils";
import { useState } from "react";
import TwoNumberEditor from "./two-number-editor";
import { CardTitle } from "../ui/card";

export const digits = (num: number) => {
  return num == 0 ? 1 : Math.floor(Math.log10(num)) + 1;
};

/**
 * get Nth (0 based) from the right
 * > getNthDigit(12345, 3)
 *  2
 * > getNthDigit(12345, 2)
 *  3
 *
 */
export function getNthDigit(
  number: number,
  pos: number,
  fromRight: boolean = true
): number {
  const array = String(number).split("");
  return Number(fromRight ? array.reverse()[pos] : array[pos]);
}

// Returns the bigger number at 0 and the smaller one at 1
export function orderNumbers(lhs: any, rhs: any) {
  const lhsDigits = lhs;
  const rhsDigits = rhs;

  return rhsDigits >= lhsDigits ? [rhs, lhs] : [lhs, rhs];
}

export default function MultiplicationSteps() {
  const [firstNumber, setFirstNumber] = useState(34);
  const [secondNumber, setSecondNumber] = useState(2312);

  const [topNumber, bottomNumber] = orderNumbers(firstNumber, secondNumber);

  const { rows } = multiplicationSteps(topNumber, bottomNumber);
  const maps: any = {
    firstNumber: {
      label: "A number to multiply",
      type: "number",
      onValueChange: setSecondNumber,
    },
    secondNumber: {
      label: "Another number to multiply",
      type: "number",
      onValueChange: setFirstNumber,
    },
  };
  
  return (
    <>
      <TwoNumberEditor
        firstNumber={firstNumber}
        secondNumber={secondNumber}
        maps={maps}
      />
      <CardTitle className="bg-gray-300 p-2 mt-2 rounded-md">Solution</CardTitle>
      <div className="flex">
        <div className="flex-grow">&nbsp;</div>
        <div className="rounded-md border-2 p-4 mt-4 text-2xl">
          <NumberWithDigits number={topNumber} />
          <NumberWithDigits number={bottomNumber} prefix="x" border="b" />

          {rows.map(({ multiplier, product, tenthPlace }, index) => {
            return (
              <div key={index}>
                {
                  <NumberWithDigits
                    number={product * Math.pow(10, tenthPlace)}
                  />
                }
              </div>
            );
          })}
          <NumberWithDigits
            number={topNumber * bottomNumber}
            border="bt"
            className="font-bold"
          />
        </div>
        <div className="flex-grow">&nbsp;</div>
      </div>
    </>
  );
}

export function NumberWithDigits({ number, prefix, border, className }: any) {
  const bottomDigits = number == 0 ? 1 : digits(number);
  const digitElements = [];
  for (var i = bottomDigits - 1; i >= 0; i--) {
    digitElements.push(
      <div key={i} className="p-1 ">
        {getNthDigit(number, i)}
      </div>
    );
  }
  const dynamicClassNames = `${
    border?.indexOf("t") >= 0 ? "border-t-red-500 border-t-[1px]" : ""
  }
  ${border?.indexOf("l") >= 0 ? "border-l-red-500 border-l-[1px]" : ""}
  ${
    border?.indexOf("b") >= 0 ? "border-b-red-500 border-b-[1px]" : ""
  } ${className}`;
  return (
    <div>
      <div className={cn("flex justify-end", dynamicClassNames)}>
        {prefix ? <div className="p-1 "> X </div> : null}
        {digitElements}
      </div>
    </div>
  );
}

function multiplicationSteps(top: number, bottom: number) {
  const rows = [];
  const bottomDigits = digits(bottom);
  for (var i = 0; i < bottomDigits; i++) {
    const thisMultiplier = getNthDigit(bottom, i);
    rows.push({
      multiplier: thisMultiplier,
      product: thisMultiplier * top,
      tenthPlace: i,
    });
  }
  return {
    top,
    bottom,
    rows,
  };
}

