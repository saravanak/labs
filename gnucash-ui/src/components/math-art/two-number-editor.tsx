import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form";
import { Input } from "@/components/ui/input";
import { zodResolver } from "@hookform/resolvers/zod";
import { useForm } from "react-hook-form";

import { z } from "zod";
export default function TwoNumberEditor({
  firstNumber,
  secondNumber,
  maps,
}: any) {
  const divideMeta = z.coerce.number().gt(0).lt(999999);

  const divisionSchema = z.object({
    firstNumber: divideMeta,
    secondNumber: divideMeta,
  });

  const form = useForm<z.infer<typeof divisionSchema>>({
    resolver: zodResolver(divisionSchema),
    mode: "onChange",
    defaultValues: {
      firstNumber,
      secondNumber,
    },
  });

  return (
    <>
      <Form {...form}>
        <form className="space-y-8">
          {Object.keys(maps).map((v: any, index: number) => (
            <FormField
              control={form.control}
              name={v}
              render={({ field }) => (
                <FormItem>
                  <FormLabel>{maps[v].label}</FormLabel>
                  <FormControl>
                    <Input
                      type={maps[v].type}
                      {...field}
                      onChange={(event) => {
                        const changedNumber = event.target.value;
                        const { data } = divideMeta.safeParse(changedNumber);
                        if (data) {
                          maps[v].onValueChange(data);
                        }
                        field.onChange(event);
                      }}
                    />
                  </FormControl>
                  <FormMessage />
                </FormItem>
              )}
            />
          ))}
        </form>
      </Form>
    </>
  );
}
