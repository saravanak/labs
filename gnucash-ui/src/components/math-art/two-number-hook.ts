import { useState } from "react";

export default function useTwoNumbers(_firstNumber: any, _secondNumber: any) {
  const [firstNumber, setFirstNumber] = useState(_firstNumber);
  const [secondNumber, setSecondNumber] = useState(_secondNumber);

  return {
    firstNumber,
    secondNumber,
    setFirstNumber,
    setSecondNumber,
  };
}
