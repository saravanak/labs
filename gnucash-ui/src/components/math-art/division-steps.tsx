import { useState } from 'react';
import { digits, getNthDigit, orderNumbers } from './multiply';

import { CardTitle } from '../ui/card';

import { clone, last } from 'lodash';

import { cn } from '@/lib/utils';
import TwoNumberEditor from './two-number-editor';
import useTwoNumbers from './two-number-hook';

function leadingNDigits(value: number, n: number) {
  return parseInt(value.toString().split('').slice(0, n).join(''));
}

function countLeadingZeroes(numberString: string) {
  const numberValue = parseInt(numberString, 10);
  return numberValue == 0
    ? numberString.length - 2
    : numberString.length - digits(numberValue);
}

/**
 *     ["100231", "096", "0042", "31"]
 *     ["0042", "0036", "00063", "1"]
 *     ["00063", "00060", "000031", ""]
 *     ["000031", "000024", "000007", ""]
 */

function singleDivisionStep(
  acc: any,
  stepindex: number,
  dividend: number,
  divisor: number
) {
  const quotient = Math.floor(dividend / divisor);
  const currentDigit = getNthDigit(quotient, stepindex, false);
  const currentDenominator = divisor * currentDigit;
  let _, x, prevRemainder, nextSeed, numeratorString, currentDenominatorString;
  if (stepindex == 0) {
    const dividentString = dividend.toString(10);
    numeratorString = dividentString.slice(0, digits(currentDenominator));

    if (parseInt(numeratorString, 10) < currentDenominator) {
      numeratorString = dividentString.slice(0, digits(currentDenominator) + 1);
    }
    nextSeed = dividentString.slice(numeratorString.length);
  } else {
    prevRemainder = (last(acc) as any)[2];
    nextSeed = (last(acc) as any)[3];
    numeratorString = prevRemainder;
  }

  currentDenominatorString = currentDenominator
    .toString(10)
    .padStart(numeratorString.length, '0');
  const currentRemainder = parseInt(numeratorString, 10) - currentDenominator;
  let currentRemainderString = currentRemainder
    .toString(10)
    .padStart(numeratorString.length, '0');

  currentRemainderString = nextSeed.length
    ? `${currentRemainderString}${nextSeed[0]}`
    : currentRemainderString;

  acc.push([
    numeratorString,
    currentDenominatorString,
    currentRemainderString,
    clone(nextSeed).slice(1),
    currentDigit,
  ]);

  if (nextSeed.length == 0) {
    return;
  } else {
    singleDivisionStep(acc, stepindex + 1, dividend, divisor);
  }
}
export function computeDivisionSteps({ dividend, divisor }: any) {
  const steps = [] as any;
  singleDivisionStep(steps, 0, dividend, divisor);
  return steps;
}

export default function DivisionSteps() {
  const twoNumberMeta = useTwoNumbers(3, 23167);

  const [dividend, divisor] = orderNumbers(
    twoNumberMeta.firstNumber,
    twoNumberMeta.secondNumber
  );

  const steps = computeDivisionSteps({ dividend, divisor });

  const quotient = Math.floor(dividend / divisor);
  const remainder = dividend % divisor;
  const divisorDigits = digits(divisor);
  const dividendDigits = digits(dividend);

  const gridColumns = dividendDigits + divisorDigits;

  const maps: any = {
    firstNumber: {
      label: 'A number to divide',
      type: 'number',
      onValueChange: twoNumberMeta.setFirstNumber,
    },
    secondNumber: {
      label: 'Another number to divide',
      type: 'number',
      onValueChange: twoNumberMeta.setSecondNumber,
    },
  };
  return (
    <>
      <TwoNumberEditor twoNumberMeta={twoNumberMeta} maps={maps} />

      <CardTitle className='bg-gray-300 p-2 mt-2 rounded-md'>
        Solution
      </CardTitle>
      <div className='grid place-content-center '>
        <div
          className={`grid font-mono text-base	tracking-[0.35ch] text-xl box-border gap-0  place-self-center `}
          style={{ gridTemplateColumns: `repeat(${gridColumns},  3ch)` }}
        >
          <Digits
            number={quotient}
            colStart={digits(divisor) + 1}
            rowStart={1}
          />
          <Digits
            number={divisor}
            colStart={1}
            rowStart={2}
            applyClasses={(index: any) => index == digits(divisor) - 1}
            borderClasses={'border-r-[0.25ch] '}
          />
          <Digits
            number={dividend}
            colStart={1 + digits(divisor)}
            rowStart={2}
            borderClasses={'border-t-[0.25ch] '}
          />
          {steps.map(
            (
              [stepDividend, multiple, remainder, nextSeed]: any,
              index: any
            ) => {
              return [stepDividend, multiple, remainder].map(
                (number, stepIndex) => {
                  let border = stepIndex == 2 ? ' border-t-[0.25ch] ' : '';

                  if (index >= 0 && stepIndex == 0) {
                    return <div className={`flex `} key={stepIndex}></div>;
                  }
                  const leadingZeroes = countLeadingZeroes(number);
                  return (
                    <>
                      {stepIndex == 2 ? (
                        <Digits
                          number={0}
                          colStart={1 + digits(divisor) + leadingZeroes - 1}
                          rowStart={2 + 4 * index + stepIndex}
                          borderClasses={border}
                        />
                      ) : null}
                      <Digits
                        number={parseInt(number)}
                        colStart={1 + digits(divisor) + leadingZeroes}
                        rowStart={2 + 4 * index + stepIndex}
                        borderClasses={border}
                      />
                    </>
                  );
                }
              );
            }
          )}
        </div>
      </div>
    </>
  );
}

function Digits({
  number,
  colStart,
  rowStart,
  borderClasses,
  applyClasses = () => true,
}: any) {
  return (
    <>
      {number
        .toString(10)
        .split('')
        .map((n: any, index: number) => {
          return (
            <div
              style={{
                gridColumnStart: colStart + index,
                gridRowStart: rowStart,
              }}
              className={cn(
                'items-center content-center aspect-square text-center  border-red-300',
                applyClasses(index) ? borderClasses : ''
              )}
              key={index}
            >
              {n}
            </div>
          );
        })}
    </>
  );
}
