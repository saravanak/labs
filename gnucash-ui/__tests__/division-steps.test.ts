import { computeDivisionSteps } from '@/components/math-art/division-steps';
import '@testing-library/jest-dom';
import { last } from 'lodash';

describe('Division steps ', () => {
  [
    // { dividend: 8, divisor: 2 },
    // { dividend: 10, divisor: 2 },
    // { dividend: 100, divisor: 2 },
    // { dividend: 101, divisor: 2 },
    { dividend: 1001, divisor: 2 },
  ].forEach(({ dividend, divisor }) => {
    it(`computes correctly ${dividend}/${divisor}`, () => {
      let computedSteps = computeDivisionSteps({ dividend, divisor });
      expect(computedSteps).toMatchSnapshot();
      expect(computedSteps.remainder).toEqual(
        last(computedSteps.divisionModel)?.remainder
      );
    });
  });
});
