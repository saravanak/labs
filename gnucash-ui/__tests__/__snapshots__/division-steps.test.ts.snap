// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Division steps  computes correctly 10/2 1`] = `
{
  "divisionModel": [
    {
      "closestMultiple": 10,
      "cumDigitsConsumed": 3,
      "currentDividend": 0,
      "currentDividendRemainingStringStr": "",
      "digitsConsumed": 2,
      "dividendPart": 10,
      "previousRemainder": null,
      "remainder": 0,
    },
  ],
  "quotient": 5,
  "remainder": 0,
}
`;

exports[`Division steps  computes correctly 100/2 1`] = `
{
  "divisionModel": [
    {
      "closestMultiple": 10,
      "cumDigitsConsumed": 3,
      "currentDividend": 0,
      "currentDividendRemainingStringStr": "0",
      "digitsConsumed": 2,
      "dividendPart": 10,
      "previousRemainder": null,
      "remainder": 0,
    },
    {
      "closestMultiple": 0,
      "cumDigitsConsumed": 4,
      "currentDividend": 0,
      "currentDividendRemainingStringStr": "",
      "digitsConsumed": 1,
      "dividendPart": 0,
      "previousRemainder": 0,
      "remainder": 0,
    },
  ],
  "quotient": 50,
  "remainder": 0,
}
`;

exports[`Division steps  computes correctly 101/2 1`] = `
{
  "divisionModel": [
    {
      "closestMultiple": 10,
      "cumDigitsConsumed": 3,
      "currentDividend": 1,
      "currentDividendRemainingStringStr": "1",
      "digitsConsumed": 2,
      "dividendPart": 10,
      "previousRemainder": null,
      "remainder": 0,
    },
    {
      "closestMultiple": 0,
      "cumDigitsConsumed": 4,
      "currentDividend": 1,
      "currentDividendRemainingStringStr": "1",
      "digitsConsumed": 1,
      "dividendPart": 1,
      "previousRemainder": 0,
      "remainder": 1,
    },
  ],
  "quotient": 50,
  "remainder": 1,
}
`;

exports[`Division steps  computes correctly 1001/2 1`] = `
{
  "divisionModel": [
    {
      "closestMultiple": 10,
      "cumDigitsConsumed": 3,
      "currentDividend": 1,
      "currentDividendRemainingStringStr": "01",
      "digitsConsumed": 2,
      "dividendPart": 10,
      "previousRemainder": null,
      "remainder": 0,
    },
    {
      "closestMultiple": 0,
      "cumDigitsConsumed": 4,
      "currentDividend": 1,
      "currentDividendRemainingStringStr": "1",
      "digitsConsumed": 1,
      "dividendPart": 0,
      "previousRemainder": 0,
      "remainder": 0,
    },
    {
      "closestMultiple": 0,
      "cumDigitsConsumed": 5,
      "currentDividend": 1,
      "currentDividendRemainingStringStr": "1",
      "digitsConsumed": 1,
      "dividendPart": 1,
      "previousRemainder": 0,
      "remainder": 1,
    },
  ],
  "quotient": 500,
  "remainder": 1,
}
`;
